cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```
@@ -91,51 +201,730 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+
+    private Cancellable notificationSubscription;
+    private ScheduledFuture<?> scheduledPolling;
+    private StreamSizeNotification lastNotification;
+    private StreamSizeNotification lastPollingInfo;
+
+    private StreamSubscriber(Id.Stream streamId) {
+      this.streamId = streamId;
+      this.scheduleTasks = Maps.newConcurrentMap();
+      this.activeTasks = new AtomicInteger(0);
+    }
+
+    public void start() throws NotificationFeedException, NotificationFeedNotFoundException {
+      notificationSubscription = notificationService.subscribe(getFeed(), this, sendPollingInfoExecutor);
+    }
+
+    @Override
+    public void cancel() {
+      if (scheduledPolling != null) {
+        scheduledPolling.cancel(true);
+      }
+      if (notificationSubscription != null) {
+        notificationSubscription.cancel();
+      }
+    }
+
+    /**
+     * Add a new scheduling task based on the data received by the stream referenced by {@code this} object.
+     * @return {@code true} if the task was created successfully, {@code false} if it already exists
+     */
+    public boolean createScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                      StreamSizeSchedule streamSizeSchedule, boolean active,
+                                      long baseRunSize, long baseRunTs, boolean persist) throws SchedulerException {
+      final StreamSizeScheduleTask newTask = new StreamSizeScheduleTask(programId, programType, streamSizeSchedule);
+      synchronized (this) {
+        StreamSizeScheduleTask previous =
+          scheduleTasks.putIfAbsent(AbstractSchedulerService.scheduleIdFor(programId, programType,
+                                                                           streamSizeSchedule.getName()),
+                                    newTask);
+        if (previous != null) {
+          // We cannot replace an existing schedule - that functionality is not wanted - yet
+          return false;
```

this should throw an exception. Or the caller should. If I understand correctly, we support updating a schedule, but through a different means.